File: crates/uic_js/js/src/runtime/dialogs.ts
// Browser dialogs over the host native (src/dialogs.rs). The promises and
// their resolvers live here; the native only queues the question and the
// host answers through __uicDialogAnswer. Installed only when the dialogs
// feature registered the native; without it the globals stay undefined.
// In a real browser these are native and synchronous, so `await confirm(…)`
// is the one spelling with identical semantics in both hosts.

export function installDialogs(): void {
    if (typeof __uic_dialog_request !== 'function') {
        return;
    }
    let nextId = 1;
    const pending = new Map<number, (value: unknown) => void>();
    (globalThis as any).__uicDialogAnswer = (id: number, value: unknown): void => {
        const resolve = pending.get(id);
        pending.delete(id);
        resolve?.(value);
    };
    const request = (kind: string, message: string, fallback: string | null): Promise<unknown> => {
        const id = nextId++;
        const answer = new Promise((resolve) => pending.set(id, resolve));
        __uic_dialog_request(id, kind, message, fallback);
        return answer;
    };
    (globalThis as any).alert = (message?: unknown): Promise<void> =>
        request('alert', String(message ?? ''), null) as Promise<void>;
    (globalThis as any).confirm = (message?: unknown): Promise<boolean> =>
        request('confirm', String(message ?? ''), null) as Promise<boolean>;
    (globalThis as any).prompt = (message?: unknown, fallback?: unknown): Promise<string | null> =>
        request(
            'prompt',
            String(message ?? ''),
            fallback == null ? '' : String(fallback),
        ) as Promise<string | null>;
}

File: crates/uic_js/src/dialogs.rs
//! Browser dialogs behind the terminal runtime: the mocked `alert`,
//! `confirm` and `prompt` queue their questions here — a thread-local,
//! because the flat natives capture nothing — and the host answers later
//! through [`JsHost::answer_dialog`](crate::JsHost::answer_dialog). The
//! promises and their resolvers stay JS-side (runtime/dialogs.ts); this
//! module is only the one-way question queue.

use std::cell::RefCell;
use std::collections::VecDeque;

/// What a dialog asks and how it answers: alert acknowledges, confirm
/// decides, prompt collects a line of text.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum DialogKind {
    Alert,
    Confirm,
    Prompt,
}

/// One question a component asked; `id` routes the answer back to the
/// promise waiting on it.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct DialogRequest {
    pub id: u32,
    pub kind: DialogKind,
    pub message: String,
    /// The prompt's prefilled text; `None` for alert and confirm.
    pub default: Option<String>,
}

thread_local! {
    static PENDING: RefCell<VecDeque<DialogRequest>> = const { RefCell::new(VecDeque::new()) };
}

/// A fresh host starts with an empty queue — questions cannot leak between
/// hosts sharing a thread (the same discipline the state slot has).
pub(crate) fn reset() {
    PENDING.with(|queue| queue.borrow_mut().clear());
}

pub(crate) fn push(request: DialogRequest) {
    PENDING.with(|queue| queue.borrow_mut().push_back(request));
}

pub(crate) fn take() -> Option<DialogRequest> {
    PENDING.with(|queue| queue.borrow_mut().pop_front())
}
